lib/graphql_metrics/analyzer.rb

```diff
@@ -0,0 +1,179 @@
+module GraphQLMetrics
+  class Analyzer < GraphQL::Analysis::AST::Analyzer
```

Comment:
I know this is WIP but a couple ideas:

* having one file is nice, but it might make things more confusing. Especially for naming?
* `Analyzer` is also a tracer right now which is a symptom of the above
* whatever this is called, it would be great to offer a single `use` method so it can act as a plugin:

```ruby
class MySchema < GraphQL::Schema
  use GraphQL::Metrics
end
```

The `use` method can handle all of this:

```ruby
instrument :query, SimpleAnalyzer
query_analyzer SimpleAnalyzer
tracer SimpleAnalyzer
```

If someone wants to opt-out of one, then they'd just do that manually by picking whatever parts they want.

Reply:
💯 !

Reply:
> having one file is nice, but it might make things more confusing. Especially for naming?

I was also thinking of the option of keeping it as one file, and just having private modules (for tracing, instrumentation methods) that get included into the analyzer class.

Reply:
Whatever makes it easier for development purposes 👍 The main thing is providing a simpler way for consumers to implement this. So at least they'd have only a single "entry point" by default.

Reply:
🤔 I might be missing something (and/or jetlagged) but the gem automatically instantiates the analyzers: https://github.com/swalkinshaw/graphql-ruby/blob/1149e07bd8b5f64f4b1bd7265d95fdeb77d6baa5/lib/graphql/analysis/ast.rb#L60

So your initializer will always be run without having to do anything manually?

Reply:
graphql-ruby does initialize analyzers by default when they are provided like so:

```ruby
  class Schema < GraphQL::Schema
    # ...
    query_analyzer MyAnalyzer
  end
```

But it doesn't do so when they are provided via `schema_defn.query_analyzer(Analyzer)`, as seen in the `self.use` example above.

Reply:
I don't think thats true since both methods just add an analyzer to the `query_analyzers` array? Either way they get instantiated later in the file I linked.

Reply:
I've push up this commit on a throwaway branch to illustrate the issue:

https://github.com/Shopify/graphql-metrics/commit/ff6f93bb74337d6baec9b75fc850015d07925b18

Reply:
Thanks @swalkinshaw for spotting the issue:

```diff
-  def self.use(schema_defn)
+  def self.use(defn)
+    schema_defn = defn.target
     schema_defn.instrument(:query, Instrumentation)
-    schema_defn.query_analyzer(Analyzer)
-    schema_defn.tracer(Tracer)
+    schema_defn.query_analyzers << Analyzer
+    schema_defn.tracers << Tracer
   end
```

Reply:
This still doesn't work unfortunately. We dug into it and it's definitely an upstream bug due how the gem handles the legacy layer under the class-based one.

Reply:
Yeah I'll repro it upstream and hope to fix it / see it fixes by the time we want to cut the 3.0 gem publicly.